exercises/dominoes/.meta/src/reference/java/Dominoes.java

```diff
@@ -0,0 +1,67 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+class Dominoes {
+	
+	List<Domino> formChain(List<Domino> inputDominoes) {
+		if (inputDominoes.size() == 0) {
+			return inputDominoes;
+		} else {
+			return formPartialChain(new ArrayList<Domino>(), inputDominoes);
+		}
+	}
+	
+	private boolean isValidChain(ArrayList<Domino> chain) {
+		return isValidPartialChain(chain) && (chain.get(0).getLeft() == chain.get(chain.size() - 1).getRight());
+	}
+	
+	private boolean isValidPartialChain(ArrayList<Domino> chain) {
+		for (int i = 0; i < chain.size() - 1; i++) {
+			if (chain.get(i).getRight() != chain.get(i+1).getLeft()) {
+				return false;
+			}
+		}
+		
+		return true;
+	}
+	
+	ArrayList<Domino> formPartialChain(ArrayList<Domino> current, List<Domino> remaining) {
+		if (remaining.size() == 0) {
+			if (isValidChain(current)) {
+				return current;
+			} else {
+				throw new ChainNotFoundException("No chain found.");
+			}
+		}
+		
+		for (int i = 0; i < remaining.size(); i++) {
+			ArrayList<Domino> newRemaining = new ArrayList<Domino>(remaining);
+			newRemaining.remove(i);
+			
+			ArrayList<Domino> newChainA = new ArrayList<Domino>(current);
+			newChainA.add(remaining.get(i));
+			
+			ArrayList<Domino> newChainB = new ArrayList<Domino>(current);
+			newChainB.add(new Domino(remaining.get(i).getRight(), remaining.get(i).getLeft()));
+			
+			if (isValidPartialChain(newChainA)) {
+				try {
+					return formPartialChain(newChainA, newRemaining);
+				} catch (ChainNotFoundException e) {
```

Comment:
Why are you catching the exception here but leaving the catch block empty? Leaving a catch block empty is generally bad practice as it can easily hide errors and make your program hard to debug. If an exception has been thrown you should need to do something to handle it or just rethrow it. There might be a special case where you're sure you need to leave the catch block empty in which case you should leave a comment explaining why this is necessary.

In your case you don't need to catch the exception at all as far as I can tell, you just want the exception to propagate up and leave it up to the caller of this method to handle or rethrow it. Does that make sense? 🙂

Reply:
I've added a comment, but I think it needs an empty catch block. It is just supposed to continue on if there is no chain found in a particular branch of the tree. Am I right?